File: examples/kotlin/micro/AuthService.kt
package micro

import java.sql.DriverManager

class AuthService {
    private val conn = DriverManager.getConnection("jdbc:sqlite:auth.db")

    fun verifyToken(token: String): String? {
        val stmt = conn.createStatement()
        val query = "SELECT user_id FROM tokens WHERE token = '$token'"
        val rs = stmt.executeQuery(query) // sink: SQL injection
        return if (rs.next()) rs.getString("user_id") else null
    }

    fun runAdminCommand(userId: String, cmd: String) {
        if (userId.isNotEmpty()) {
            Runtime.getRuntime().exec("notify-admin $cmd") // sink: command injection
        }
    }
}

File: examples/typescript/micro/auth_service.ts
import { execSync } from "child_process";
import mysql from "mysql";

const db = mysql.createConnection({ host: "localhost", database: "auth" });

export function verifyToken(token: string): string | null {
  // Inline the SQL string at the call site so the rule's
  // arg-shape SQL-keyword constraint can fire syntactically.
  db.query("SELECT user_id FROM tokens WHERE token = '" + token + "'");  // sink: SQL injection
  return token;
}

export function runAdminCommand(userId: string, cmd: string): void {
  if (userId) {
    execSync("notify-admin " + cmd);  // sink: command injection
  }
}
